### PR TITLE
pop-os-shell: init at 1.1.0

### PR DESCRIPTION
**Pull request by remunds**
###### Motivation for this change

pop-os-shell version 1.0.0 was released 20 days ago. Now it's time to be a package :)
This is my first pr, so please tell me, if something is wrong :)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{ stdenv, fetchFromGitHub, nodePackages, glib }:
+
+stdenv.mkDerivation rec {
+  pname = "pop-os-shell";
+  version = "1.0";
```
Comment (SuperSandro2000) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:5]: If you set this to 1.0.0 you can reuse it in fetchFromGitHub.

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    owner = "pop-os";
+    repo = "shell";
+    rev =  "1.0.0";
+    sha256 = "sha256-gv9knhG1/HOUsIyPWfBRBpJWg/bfRXbbg/x2QDgSmls=";
+  };
+
+  buildInputs = [ glib nodePackages.typescript ];
+
+
+  uuid = "pop-shell@system76.com";
```
Comment (SuperSandro2000) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:17]: is this an env that is required for something?

Comment (remunds) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:17]: I'm not sure. Just saw this being set in all the other extensions. 
Should i remove it?

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -7289,6 +7289,12 @@
     githubId = 801525;
     name = "rembo10";
   };
+  remunds = {
+    email = "raban.emunds@gmail.com";
+    github = "remunds";
+    githubId = 61607050;
+    name = "R. Emunds";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix:7297]: Please make this into a separate commit:

```
maintainers: add <name>
```

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix (hunk 2 above)
Comment (jonringer) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:17]: I would remove it if it's not required

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  makeFlags = [ "INSTALLBASE=$(out)/share/gnome-shell/extensions" ];
+
+  meta = with stdenv.lib; {
+    description = "Keyboard-driven layer for GNOME Shell which allows for quick and sensible navigation and management of windows";
```
Comment (jonringer) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:22]: Preference for <80 characters for description. If you would like to have an extended description, please use `longDescription`
```suggestion
    description = "Keyboard-driven layer for GNOME Shell";
```

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{ stdenv, fetchFromGitHub, nodePackages, glib }:
+
+stdenv.mkDerivation rec {
+  pname = "pop-os-shell";
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "pop-os";
+    repo = "shell";
+    rev =  "${version}";
```
Comment (jonringer) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:10]: no need for interpolation, it's already a string.
```suggestion
    rev = version;
```

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    sha256 = "sha256-gv9knhG1/HOUsIyPWfBRBpJWg/bfRXbbg/x2QDgSmls=";
+  };
+
+  buildInputs = [ glib nodePackages.typescript ];
+
```
Comment (jonringer) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:15]: extra newline
```suggestion
```

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "pop-os";
+    repo = "shell";
+    rev =  "${version}";
+    sha256 = "sha256-gv9knhG1/HOUsIyPWfBRBpJWg/bfRXbbg/x2QDgSmls=";
+  };
+
+  buildInputs = [ glib nodePackages.typescript ];
```
Comment (jonringer) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:14]: these are likely just need for glibschemas and running ts
```suggestion
  nativeBuildInputs = [ glib nodePackages.typescript ];
```

**Comment by remunds** [comment]
Should be good now :+1:

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    sha256 = "sha256-gv9knhG1/HOUsIyPWfBRBpJWg/bfRXbbg/x2QDgSmls=";
+  };
+
+  nativeBuildInputs = [ glib nodePackages.typescript ];
+
+  makeFlags = [ "INSTALLBASE=$(out)/share/gnome-shell/extensions" ];
+
+  meta = with stdenv.lib; {
+    description = "Keyboard-driven layer for GNOME Shell";
+    license = licenses.gpl3;
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: `gpl3` is unclear, use `gpl3Plus` or `gpl3Only`. See https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206 for more details.

**Comment by jtojnar** [comment]
cc https://github.com/NixOS/nixpkgs/issues/92769 (@mogorman @Technical27 @operator-name)

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix (hunk 8 above)
Comment (remunds) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: I'm not entirely sure, which version to use. The license can be found here: https://github.com/pop-os/shell/blob/master/LICENSE. Also on the web i could not really identify what the difference is between gpl3-only and gpl3+. Can you tell me what we have here? @jtojnar

Comment (jtojnar) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: Usually, the precise license is listed in README, but that is not the case [here](https://github.com/pop-os/shell#license). Try asking on the upstream issue tracker.

Comment (mmstick) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: The license is [here](https://github.com/pop-os/shell/blob/master/LICENSE)

Comment (jtojnar) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: That is just license terms, it needs to be explicitly specified whether the copyright owner allows using latter versions of the license, see the GNU links in the [Discourse threads](https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206).

**Comment by buckley310** [comment]
This looks like it might be a duplicate of https://github.com/NixOS/nixpkgs/pull/102150

**Comment by jtojnar** [comment]
Oh, right. I was certain that we already had a PR open but could not find it with GitHub search :woman_shrugging:

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  src = fetchFromGitHub {
+    owner = "pop-os";
+    repo = "shell";
+    rev =  version;
+    sha256 = "sha256-gv9knhG1/HOUsIyPWfBRBpJWg/bfRXbbg/x2QDgSmls=";
+  };
+
+  nativeBuildInputs = [ glib nodePackages.typescript ];
+
+  makeFlags = [ "INSTALLBASE=$(out)/share/gnome-shell/extensions" ];
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:16]: This one is nicer in that it does not abuse `DESTDIR`.

**Comment by maxeaubrey** [comment]
Oh, neat to see someone get around to packaging this! I've had a commit sitting around waiting for a 1.0 release that might help a bit - I see this doesn't install the glib schema or set the uuid like other gnome extensions.

Also, it might be nice to also package `pop-shell-shortcuts` (https://github.com/pop-os/shell-shortcuts) to go with it, I believe otherwise the extension will try to open a website with the shortcuts.

Here's my commit for `pop-shell`: https://github.com/maxeaubrey/nixpkgs/commit/e46df8ddff5facaf57b86e5e7d5be189000e15a1
And here's `pop-shell-shortcuts`: https://github.com/maxeaubrey/nixpkgs/commit/5c14f15e6a2bfc9b6a11457ccb048a3668f10259

**Comment by jtojnar** [comment]
Looking at the source code, it should install schemas:

https://github.com/pop-os/shell/blob/1.0.0/Makefile#L74
https://github.com/pop-os/shell/blob/1.0.0/Makefile#L43

**Comment by remunds** [comment]
@maxeaubrey I had uuid set first, but i was advised to remove it if it is unnecessary.
Regarding pop-shell-shortcuts: I haven't had the problem with the website opening up and currently am using the package without shell-shortcuts. But maybe it's a good thing. Maybe putting it up as a second package is a good idea.

**Comment by maxeaubrey** [comment]
@remunds Oh, I guess uuid is obsolete now? 🤷‍♀️ Fair enough.

@jtojnar I had problems with the schema being missing and the extension not working correctly otherwise, but I cannot reproduce right now. Soz, what I said can be ignored on that.

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix (hunk 8 above)
Comment (ymarkus) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: IIRC if not listed explicitly it should be GPL3 only, shouldn't it?

Comment (jtojnar) [reply pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: Yes, that is all that can be inferred without any more info. Though people often intend `-or-later` so it does not hurt to ask.

**Comment by remunds** [comment]
As i did not get an answer on my issue (https://github.com/pop-os/shell/issues/738), i just changed the license to gpl3Only. Hopefully that is fine. I also realized that there is a version 1.1.0 now, so i also updated the derivation. Lastly, i rebased to the current master branch.
Is there anything else missing?

**Comment by jtojnar** [comment]
As long as someone verifies that it works, looks good to me.

**Comment by ymarkus** [comment]
I tried it and everything works except for the button "floating window exceptions" and "Active Hint Color". This is the error message:
```
JS ERROR: GLib.SpawnError: Failed to execute child process “gjs” (No such file or directory)
                                                  color_selector/<@/run/current-system/sw/share/gnome-shell/extensions/pop-shell@system76.com/panel_settings.js:236:25
```
and
```
JS ERROR: GLib.SpawnError: Failed to execute child process “gjs” (No such file or directory)
                                                  async_process@/run/current-system/sw/share/gnome-shell/extensions/pop-shell@system76.com/utils.js:76:10
                                                  exception_dialog@/run/current-system/sw/share/gnome-shell/extensions/pop-shell@system76.com/extension.js:313:15
                                                  floating_window_exceptions/<@/run/current-system/sw/share/gnome-shell/extensions/pop-shell@system76.com/panel_settings.js:80:13
                                                  activate@resource:///org/gnome/shell/ui/popupMenu.js:191:14
                                                  vfunc_button_release_event@resource:///org/gnome/shell/ui/popupMenu.js:138:14
```

**Comment by jtojnar** [comment]
That means we need to patch those files with `gjs` paths, see https://github.com/NixOS/nixpkgs/blob/2a058487cb7a50e7650f1657ee0151a19c59ec3b/pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix#L16-L21 for an example.

**Comment by remunds** [comment]
As I am not using those features, I did not find the bugs. Unfortunately, I have never applied a patch before and do not really know how to patch gjs files (and also which files need to be patched). Does anyone have an Idea on how to do that or can help me with it?

**Comment by jtojnar** [comment]
You would clone the pop-os repo, find the gjs references (see the files in https://github.com/NixOS/nixpkgs/pull/104160#issuecomment-749005632), replace them with `@gjs@/bin/gjs` and then create a patch using `git diff > fix-paths.patch`. You can then apply it as shown in https://github.com/NixOS/nixpkgs/pull/104160#issuecomment-749008197.

**Comment by mmstick** [comment]
I would recommend submitting patches upstream for locating the system path if the default GLib subprocess launcher can't find the binaries in your system path. You may add additional search paths with a subprocess launcher. There are a couple areas in Pop Shell that call out gjs, including some of the default plugins which are gjs scripts.

**Comment by jtojnar** [comment]
@mmstick The point of Nix is that there is not system path – every package is self-contained and hardcodes paths to everything it needs (in theory anyway). For some things like script shebangs, we have tooling that hardcodes the correct path but for `exec` calls in code, a more manual approach is required.

**Comment by mmstick** [comment]
The code that manually calls upon `gjs` can be changed to make use of the shebang line.

**Comment by jtojnar** [comment]
Right, that would be better for us too, we did the same with [GSConnect](https://github.com/GSConnect/gnome-shell-extension-gsconnect/commit/71c86dbb1e3434859331de56bbf0bb38560c8baa). @remunds are you able to submit such patch upstream?

**Comment by ymarkus** [comment]
I've found another issue, although I'm not sure if it's maybe only on my end:
Deactivating "Show window titles" doesn't hide title bars.

**Comment by benneti** [comment]
This one is probably due to using Wayland where this is not supported upstream.

**Comment by remunds** [comment]
@jtojnar I will have a look at that later that day, but I am not sure whether I am capable enough. I'll let you know. Do I understand it right that the current approach is to create a patch that puts the shebang #!/usr/bin/env gjs in every file that needs it and then somehow modify the way those files are called?

@ymarkus that works for me.

**Comment by jtojnar** [comment]
@remunds Yes, that is correct. The files need to be invoked as scripts. Also, for that to work, you will need to make them executable if they are not already.

**Comment by remunds** [comment]
I'll give that a try, thanks :)

**Comment by remunds** [comment]
I was able to get the "floating window exception"-window with the other approach though.

**Comment by remunds** [comment]
Now its just the active hint color that doesn't work. Does anyone have an idea?

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/fix-gjs.patch [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+diff --git a/src/panel_settings.ts b/src/panel_settings.ts
+index 9ee9c68..8edea9f 100644
+--- a/src/panel_settings.ts
++++ b/src/panel_settings.ts
+@@ -323,7 +323,7 @@ function color_selector(ext: Ext, menu: any) {
+     color_selector_item.add_child(color_button);
+     color_button.connect('button-press-event', () => {
+         let path = Me.dir.get_path() + "/color_dialog/main.js";
+-        let resp = GLib.spawn_command_line_async(`gjs ${path}`);
++        let resp = GLib.spawn_command_line_async(`exec .${path}`);
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/fix-gjs.patch:57]: You need the same here:

```suggestion
+        let resp = GLib.spawn_command_line_async(path);
```

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    sha256 = "sha256-KOp/0R7P/iH52njr7JPDKd4fAoN88d/pfou2gWy5QPk=";
+  };
+
+  nativeBuildInputs = [ glib nodePackages.typescript ];
+
+  patches = [
+    (substituteAll {
+      src = ./fix-gjs.patch;
+      inherit gjs;
+    })
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:20]: You are not actually substituting anything so you can just use.

File: pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+  src = fetchFromGitHub {
+    owner = "pop-os";
+    repo = "shell";
+    rev = version;
+    sha256 = "sha256-KOp/0R7P/iH52njr7JPDKd4fAoN88d/pfou2gWy5QPk=";
+  };
+
+  nativeBuildInputs = [ glib nodePackages.typescript ];
+
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/extensions/pop-os-shell/default.nix:15]: You will still need to add `buildInputs = [ gjs ];` so that `patchShebangs` can find it and substitute it.

**Comment by hmenke** [comment]
I've been using this for a few weeks now (updated to 1.2.0 though) on NixOS 20.09 under GNOME 3.36.5 Wayland and apart from the known upstream Wayland wonkiness, this actually works quite nicely.

**Comment by hpfr** [comment]
Tangentially related in case anyone is interested: https://blog.system76.com/post/648371526931038208/cosmic-to-arrive-in-june-release-of-popos-2104

https://github.com/pop-os/cosmic

It has pop-shell as a dependency.

**Comment by genofire** [comment]
Solved in  #147542 - i believe we could close this PR now.
Maybe you add yourself to maintainers ;)
